src/Step/Deploy/InstallUpdate/ConfigUpdate/DbConnection.php

```diff
@@ -60,83 +64,322 @@ class DbConnection implements StepInterface
+     * In the case when the database was split with the user configuration then sets the flag '.ignore_split_db'
+     * If the flag '.ignore_split_db' exists, the split process will be ignored
+     *
+     * @throws FileSystemException
+     * @throws ConfigException
+     * @throws ConfigurationMismatchException
      */
     public function execute()
     {
-        $config = $this->configReader->read();
+        $dbConfig = $this->getDbConfigData();
+
+        if (!isset($dbConfig[DbConfig::KEY_CONNECTION])) {
+            /**
+             * Is calling only in case when database relationship configuration doesn't exist and
+             * database is not configured through .magento.env.yaml or env variable.
+             * It's workaround for scenarios when magento was installed by raw setup:install command
+             * not by deploy scripts.
+             */
+            $this->logger->notice(
+                'Database relationship configuration doesn\'t exist'
+                . ' and database is not configured through .magento.env.yaml or env variable.'
+                . ' Will be applied the previous database configuration.'
+            );
+            return;
+        }
 
         $this->logger->info('Updating env.php DB connection configuration.');
-        $config['db'] = $this->mergedConfig->get();
-        $config['resource'] = $this->resourceConfig->get();
 
-        $this->addLoggingAboutSlaveConnection();
-        $this->configWriter->create($config);
+        $mageConfigDbConnections = $this->getMageConfigData()[DbConfig::KEY_DB][DbConfig::KEY_CONNECTION];
+        $mageSplitConnectionsConfig = array_intersect_key(
+            $mageConfigDbConnections,
+            array_flip(DbConfig::SPLIT_CONNECTIONS)
+        );
+
+        $isCustomDefaultConnection = !$this->isSameConnection(
+            $dbConfig[DbConfig::KEY_CONNECTION][DbConfig::CONNECTION_DEFAULT],
+            $mageConfigDbConnections[DbConfig::CONNECTION_DEFAULT]
+        );
+
+        if (!empty($mageSplitConnectionsConfig) && $isCustomDefaultConnection) {
+            $this->logger->notice(
+                'Database was already split but deploy was configured with new connection.'
+                . ' The previous connection data will be ignored.'
+            );
+        }
+
+        $useSlave = $this->stageConfig->get(DeployInterface::VAR_MYSQL_USE_SLAVE_CONNECTION);
+        $slaveIsAvailable = isset($dbConfig[DbConfig::KEY_SLAVE_CONNECTION]);
+
+        if (empty($mageSplitConnectionsConfig)
+            || (!empty($mageSplitConnectionsConfig) && $isCustomDefaultConnection)) {
+            $this->updateMainConnectionsConfig($useSlave, $slaveIsAvailable);
+            return;
+        }
+
+        $customSplitConnections = $this->getDifferentConnections(
+            $mageSplitConnectionsConfig,
+            $dbConfig[DbConfig::KEY_CONNECTION]
+        );
+
+        if (!empty($customSplitConnections)) {
```

Comment:
We need to check if FLAG_IGNORE_SPLIT_DB exists and remove it if so.
This flag can be left from the previous deployment in case when some error happened between `db-connection` and `split-db` steps.
We can do the same as we do with `deploy_is_failed` flag: remove it at the beginning of the deploy phase.

Reply:
Deletion of this flag was added to src/Step/Deploy/RemoveDeployFailedFlag.php